redis_shard.go

```diff
@@ -383,52 +337,48 @@ func (s *RedisShard) runDataPipeline() error {
 					break loop
 				}
 			}
-
-			conn := s.pool.Get()
-
+		}
+		for _, dr := range drs {
+			if dr.script != nil {
+				_ = pipe.EvalSha(context.Background(), dr.script.Hash(), dr.keys, dr.args...)
+			} else {
+				_ = pipe.Do(context.Background(), dr.args...)
+			}
+		}
+		cmds, err := pipe.Exec(context.Background())
+		if err == redis.Nil {
```

Comment:
It seems we can miss the error as `go-redis` always returns first non-nil error when executing Pipeline and I have not found it checks for redis.Nil internally.

Reply:
This `err == redis.Nil` is something I don't kown how to handle it correctly. For Redis scripts that returns nothing, like `addPresenceSource`, Redis will reply with a nil string, `go-redis` takes it as a error and returns a `redis.Nil` error.

If we don't ignore the `redis.Nil` here, then the whole `runDataPipeline` will exit and rerun, that's a huge performace loose.

Reply:
> If we don't ignore the redis.Nil here, then the whole runDataPipeline will exit and rerun, that's a huge performance loss.

Yes, definitely we should not restart pipeline. For me it seems like an issue of `go-redis` - because it seems that `redis.Nil` is not actually an error so `pipe.Exec` should not return it as an `error` at all.

Possibly we should just ignore `error` from `pipe.Exec` and just iterate over returned `cmds` to handle each error individually 🤔

Reply:
Pushed a commit that ignores errors from `pipe.Exec`, and defer the `redis.Nil` judgment to the upper lever (`Presencemanager.addPresence` and `removePresence`) where it knows whether `redis.Nil` is expected or not.

Reply:
Cool, so this should also effectively solve the concerns in [this comment](https://github.com/centrifugal/centrifuge/pull/235#discussion_r966697332) right?

Reply:
I think so, current behavior is in line with this statement:
> we can change Centrifuge code a bit to iterate over pipeline Exec result and manually extract which commands were successful and which failed. And extracting cmd results from successful only.